### PR TITLE
KAFKA-8523 Enabling InsertField transform to be used with tombstone events

### PR DESCRIPTION
**Pull request by gunnarmorling**
Fixes https://issues.apache.org/jira/browse/KAFKA-8523

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by gunnarmorling** [comment]
Hey @rhauch, a minor fix to the `InsertField`SMT -- Feedback welcome!

**Comment by ftardif** [comment]
So, does an insert config on a tombstone <k,null> will still produce a <k, null> (hence ignored), or will it produce a <k, v> where v contains only the inserted fields. If it does not propagate the null value, I am concerned it will skew the elasticsearch connector config `behavior.on.null.values` 
that can be used to explicitly delete index entries on null.

image

File: connect/transforms/src/main/java/org/apache/kafka/connect/transforms/InsertField.java
Comment (rhauch) [inline connect/transforms/src/main/java/org/apache/kafka/connect/transforms/InsertField.java]: Nit: let's avoid moving all of these imports around. Our convention is to place the `java` and static imports at the end, and moving them unnecessarily just complicates maintenance.

File: connect/transforms/src/test/java/org/apache/kafka/connect/transforms/InsertFieldTest.java
Comment (rhauch) [inline connect/transforms/src/test/java/org/apache/kafka/connect/transforms/InsertFieldTest.java]: Nit: might be a bit clearer to add "WithoutSchema":
```suggestion
    public void insertConfiguredFieldsIntoTombstoneEventWithoutSchema() {
```

**Review by rhauch — COMMENTED** [review]
Thanks, @gunnarmorling! This looks great, but I have two minor suggestions.

**Comment by ftardif** [comment]
before merging this PR, can we discuss the impact of applying transformation on null values? I strongly believe we should rather propagate the null values untouched even when transformations are configured. 
 
see: https://issues.apache.org/jira/browse/KAFKA-8523?focusedCommentId=16931777&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-16931777

**Review by rhauch — CHANGES_REQUESTED** [review]
@ftardif, please see the discussion on https://issues.apache.org/jira/browse/KAFKA-8523. I think @gunnarmorling, you, and I all agree that this PR should be changed to pass through and not modify tombstone records.

This PR has not yet been approved, but just to be safe I've marked this PR as `Request changes`.

**Comment by ftardif** [comment]
@gunnarmorling are you planning to provide an adjusted PR soon in order to start testing this functionality?

**Comment by gunnarmorling** [comment]
Yes, I'll try and send something in the course of the week. Thanks for the
reminder!

>

File: connect/transforms/src/main/java/org/apache/kafka/connect/transforms/InsertField.java
Comment (gunnarmorling) [reply connect/transforms/src/main/java/org/apache/kafka/connect/transforms/InsertField.java]: Reverted these.

File: connect/transforms/src/test/java/org/apache/kafka/connect/transforms/InsertFieldTest.java
Comment (gunnarmorling) [reply connect/transforms/src/test/java/org/apache/kafka/connect/transforms/InsertFieldTest.java]: Done.

**Comment by gunnarmorling** [comment]
@rhauch, @ftardif, I've updated the PR and force-pushed as per the discussion on the JIRA issue; tombstones are returned unmodified now by the `InsertField` transformation. I have also addressed the other review remarks regarding imports and method naming.

**Comment by ftardif** [comment]
thanks @gunnarmorling . What are the next steps to get that PR merged?

**Comment by gunnarmorling** [comment]
Good question; I suppose approval by @rhauch or any other committer should do the trick.

**Review by rhauch — COMMENTED** [review]
LGTM. Thanks, @gunnarmorling!

I plan to merge shortly.

**Comment by gunnarmorling** [comment]
Excellent! To be sure, would that release be 2.4? Is there a way to learn
about the timelines for code freezes for future reference? Thanks again!

>

**Comment by rhauch** [comment]
@gunnarmorling yes, I hope to get this into the AK 2.4.0 release. For information about release plans, checkout the dev list email thread with subject `[DISCUSS] Apache Kafka 2.4.0 release` (or the specific release you're interested in), which points to the [AK 2.4.0 release plan](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=125307901).

**Review by rhauch — APPROVED** [review]
LGTM.

**Comment by rhauch** [comment]
Merged back as far back as the 1.0 branch. See https://issues.apache.org/jira/browse/KAFKA-8523 for details.

**Comment by gunnarmorling** [comment]
Excellent; thanks for merging and backporting, Randall!

>

**Pull request merged**
